doc/developer/design/20231201_catalog_migration_to_persist.md

```diff
@@ -0,0 +1,164 @@
+transitions are harder to reason about.
+
+### Migrate From Stash to Persist
+
+Below are the steps to migrate from the stash to persist.
+
+1. Open the stash catalog.
+2. Open the persist catalog.
+3. Set the stash tombstone to `true`.
+4. If the persist tombstone is `false`, then we're done.
+5. Read stash catalog snapshot.
+6. Write catalog snapshot to persist.
+7. Set the persist tombstone to `false`.
+
+NOTE: Steps (6) and (7) can be done as a single compare and append operation as a performance
+optimization.
+
+### Rollback From Persist to Stash
+
+Below are the steps to rollback from persist to the stash.
+
+1. Open the stash catalog.
+2. Open the persist catalog.
+3. Set the stash tombstone to `false`.
+4. If the persist tombstone is `true` or doesn't exist, then we're done.
+5. Read persist catalog snapshot.
+6. Write catalog snapshot to stash.
+7. Set the persist tombstone to `true`.
+
+### States
+
+Below is a table describing all possible states of (stash_tombstone, persist_tombstone) and what
+they
+indicate about where the source of truth catalog data is located.
+
+| Stash Tombstone | Persist Tombstone | Source of Truth | Explanation                           |
+|-----------------|-------------------|-----------------|---------------------------------------|
+| None            | _                 | Stash           | Migration has never been initiated.   |
+| _               | None              | Stash           | Migration has never fully completed.  |
+| Some(true)      | Some(false)       | Persist         | Migration has completed successfully. |
+| Some(false)     | Some(true)        | Stash           | Rollback has completed successfully.  |
+| Some(true)      | Some(true)        | Stash           | Migration crashed before completing.  |
+| Some(false)     | Some(false)       | Persist         | Rollback crashed before completing.   |
+
+### State Transitions
+
+Below is a state transition diagram that shows all possible states of
+(stash_tombstone, persist_tombstone) and how the states transition. Each node has the format of
+(stash_tombstone, persist_tombstone). Each edge indicates what steps from the algorithm above
+triggers that specific state transition. Next to each node is either `Stash` if the stash is the
+source of truth, or `Persist` if persist is the source of truth.
+
+![state-transitions](./static/catalog_migration_to_persist/catalog_migration_multi_tombstone.png)
+
+## Open questions
+
+- There are certain classes of optimizations we can make to the startup process depending on the
+  initial state. For example, if we are migrating from the stash to persist and the stash has never
+  been initialized, then don't initialize the stash. Another example is that if we're rolling back
+  from persist to the stash and there's data format migrations, then we should skip migrations in
+  persist. Should we make these optimizations? My opinion is no because they will complicate the
```

Comment:
I also think no. Reducing complexity and the potential for bugs here is more important than minor startup time optimizations